src/VisualStudio/Core/Def/ProjectSystem/BrokeredService/WorkspaceProject.cs

```diff
@@ -0,0 +1,146 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Remote.ProjectSystem;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.ProjectSystem.BrokeredService
+{
+    internal sealed class WorkspaceProject : IWorkspaceProject
+    {
+        // For the sake of the in-proc implementation here, we're going to build this atop IWorkspaceProjectContext so semantics are preserved
+        // for a few edge cases. Once the project system has moved onto this directly, we can flatten the implementations out.
+        private readonly IWorkspaceProjectContext _project;
+
+        public WorkspaceProject(IWorkspaceProjectContext project)
+        {
+            _project = project;
+        }
+
+        public ValueTask DisposeAsync()
+        {
+            _project.Dispose();
+            return ValueTaskFactory.CompletedTask;
+        }
+
+        public async Task AddAdditionalFilesAsync(IReadOnlyCollection<string> additionalFilePaths, CancellationToken cancellationToken)
```

Comment:
random thought, but what are the cancellation semantics here?  this partially doesn't matter (since you never use it), but are tehre cases where we would cancel midway through adding things?

Reply:
@CyrusNajmabadi We're adding cancellation tokens to follow the RPC recommended design, but we wouldn't cancel mid-way at all.

This was one question I had for @AArnott was how cancellation is expected to be handled on mutating methods like this; although StreamJsonRpc could respect the cancellation before the call ever gets to the server in the first place, once it's sent over the wire the client will know whether we did any cancellation on the server.